man/man1/cdf2soil.1
.TH cdf2soil 1  "USER COMMANDS"
.SH NAME
cdf2soil \- convert NetCDF file into binary file
.SH SYNOPSIS
.B cdf2soil
[\-var name] [\-{float|double}] [\-scale s] [\-soilmap]
.I netcdffile gridfile soilfile
.SH DESCRIPTION
Program converts NetCDF input data into soil code or binary input data and writes grid file.
.SH OPTIONS
.TP
\-var name
Variable name in NetCDF file, default is 'stexture'.
.TP
\-scale factor
Scaling factor for clm grid file. Default is 0.01.
.TP
\-{float|double}
Write float or double values in CLM grid file without applying the scaling factor, default is short.
.TP
\-soilmap
Use built-in soil map.
.TP
.I netcdffile
Filename of NetCDF file converted.
.TP
.I gridfile
Filename of grid data file written.
.TP
.I soilfile
Filename of soil code data file written.
.SH EXAMPLES
.B cdf2soil
-var soilcode soilcode.nc grid.clm soilcode.bin
.PP
.SH EXIT STATUS
.B cdf2soil
returns a zero exit status if the soil code file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
cdf2coord(1), cdf2clm(1), cdf2bin(1), clm2cdf(1), bin2cdf(1), country2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security


## [Unreleased]


## [6.1.4] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Fabian Stenzel (stenzel@pik-potsdam.de), David Hötten (davidho@pik-potsdam.de)

### Added

- Utility `cmprestart` added to compare two restart files and check for identity.
- Function `bstruct_skiparray()` and corresponding test `test_skiparray.c` added in `bstruct` library to skip arrary elements.
- Function `bstruct_gettype()` added to `bstruct` library to get datatype of object in restart file.
- Function `bstrucht_printnamestack()` added for more comprehensive error messages.
- Comparison of restart files with `cmprestart` added to merge request template.

### Changed

- Index array object is always skipped in `bstruct_findobject.c`, test `test_indexarray.c` updated to test for reading array without seeking to an index array.
- Check for seeking to position within file added to `bstruct_findobject.c`, `bstruct_open.c`, `bstruct_wopen.c` and `bstruct_seekindexarray.c`.
- Constant `MAXLEVEL` renamed to `BSTRUCT_MAXLEVEL`.
- Macros `getname()` and `isinvalidtoken()` renamed to `bstruct_getname()` and `bstruct_isinvalidtoken()`.
- String comparison in `bstruct_fprintdata` replaced by call to `isnan()` and `isinf()`.
- New token `bstruct_dzero` defined for restart files to distinguish between float and double zeros, version of bstruct files changed to 2.


## [6.1.3] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth(wirth@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jannes Breier (breier@pik-potsdam.de)

### Added

- Option `-metafile` and `-json` added to `cutclm` and `catclm`.
- Option `-double` added to `cdf2soil.c` to write grid file with double datatype.

### Changed

- New datatype `Metadata` defined in `types.h` combining maps, attributes, basetemp and hlimit arrays used by `opendata()`.
- basetemp, countrymap and hlimit arrays are copied from JSON metafiles to newly created metafiles in `regridclm`, `cutlcm`, `catclm`, and `mathclm`.
- Scaling to sum of stand fractions removed for output `"vegc_avg"`.
- Utility `getcountry` reads now JSON metafile for country codes including the alpha-3 country codes and global grid instead of CLM file.
- If `-pedantic` option is set, non-matching titles in input files stops `lpjml` with an error.
- Datatype of `lon`, `lat`, and `area` elements of `Coord` changed from `Real` to `double`.
- Missing path to grid file added in `openmetafile.c`.

### Fixed

- Warning for float coordinates corrected in `cdf2coord.c`, `cdf2grid.c` and `cdf2soil.c`.
- Variables initialized to zero in `reservoir2cdf.c` and `restart2yaml.c` to avoid warnings by `gcc`.
- Carbon balance fixed in `reclaim_land.c`.
- Check for NULL pointer added in `getuser.c` to avoid SEGV if user name cannot be resolved.
- Verbose output corrected for input without time axis in `cdf2clm.c`.
- Output is rescaled by the actual sum of stand fractions in `update_daily_cell.c` and `fwriteoutput.c` to avoid division by zero (issue #411).
- Default timestep set to 1 in `regridclm.c`.
- Variable `depth` initialized to zero before layer loop in `fwriteoutput.c` to correct `soilc_1m` output.
- Length of buffer for hostname corrected in `gethost.c`.
- Output of unprintable characters corrected in `fputprintable.c`.
- Memory leak fixed in `openmetafile.c` in case of error.


## [6.1.2] - 2026-06-16

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Fixed

- Check for last year for lightning and human ignition disabled in `filesexist.c`, `lpjcheck` does not return an error for standard input.


## [6.1.1] - 2026-06-12

### Contributors

author: Jannes Breier (breier@pik-potsdam.de),  Werner von Bloh (bloh@pik-potsdam.de)
code review: Christoph Müller (cmueller@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)

### Added

- `logo.png` added to `man` directory
- `LICENSES` folder
- `LICENSES/CC0-1.0.md` added for artwork license

### Changed

- `AUTHORS` renamed to `AUTHORS.md` and restructured for active and previous authors
- `INSTALL` renamed to `INSTALL.md` and adjusted
- `LICENSE` renamed to `LICENSES/AGPL-3.0.md` and moved to `LICENSES` folder
- `README` renamed to `README.md` and expanded
- `configure.sh`, `src/lpj/copyright.c` and `src/lpj/printlicense.c` adjusted to reflect renaming of files
- Updated module version for `oneAPI` in `run_simulate_default.sh`


## [6.1.0] - 2026-06-12

### Contributors

- author: Luke Oberhagemann (lukeober@pik-potsdam.de), Markus Drüke (Markus.Drueke@dwd.de), Maik Billing (billing@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Option `-inpath` added to `configure.sh` to set path for input files.
- Function `fprintclimbuf()` added in `climbuf.c` to print variables and is called in `fprintcell.c`.
- Function `fprintbuffer()` added in `buffer.c` to print variables and is called in `fprintclimbuf.c`.
- Lakes setting stored in restart file and setting compared to actual setting in `openrestart.c`.

src/utils/cdf2soil.c
/**************************************************************************************/
/**                                                                                \n**/
/**                c  d  f  2  s  o  i  l  .  c                                    \n**/
/**                                                                                \n**/
/**     Converts NetCDF data file into LPJ soil and coord data file                \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#ifdef USE_NETCDF
#include <netcdf.h>
#endif

#define USAGE  "Usage: %s [-var name] [-{float|double]] [-scale s] [-soilmap] netcdffile coordfile soilfile\n"

#define error(rc) if(rc){ fprintf(stderr,"ERROR403: Cannot read '%s': %s.\n",argv[i],nc_strerror(rc)); return EXIT_FAILURE;}

#define SOIL_NAME "stexture"

int main(int argc,char **argv)
{
#ifdef USE_NETCDF
  int rc,ncid,var_id,i,ndims,dimids[2],latlon_id;
  char name[NC_MAX_NAME+1];
  size_t ilat,ilon;
  double *lat,*lon;
  size_t lat_len,lon_len;
  int missing_value;  /**< missing/fill value in file */
  int *data;
  Header header;
  double scalar=0.01;
  FILE *out,*soil;
  Byte soilcode;
  Bool ismap,scalar_set;
  Byte soilmap[]={Sa,LoSa,SaLo,SiLo,Lo,SaClLo,SiClLo,ClLo,SaCl,SiCl,Cl,Si};
  Coord coord;
  char *var=SOIL_NAME;
  char *endptr;
  header.scalar=0.01;
  header.datatype=LPJ_SHORT;
  scalar=0.01;
  ismap=scalar_set=FALSE;
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-var"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-var'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        var=argv[++i];
      }
      else if(!strcmp(argv[i],"-float"))
      {
        header.datatype=LPJ_FLOAT;
        header.scalar=1;
        scalar=1;
      }
      else if(!strcmp(argv[i],"-double"))
      {
        header.datatype=LPJ_DOUBLE;
        header.scalar=1;
        scalar=1;
      }
      else if(!strcmp(argv[i],"-soilmap"))
        ismap=TRUE;
      else if(!strcmp(argv[i],"-scale"))
      {
        if(argc==i+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-scale'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        scalar=strtod(argv[++i],&endptr);
        header.scalar=scalar;
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for scale.\n",argv[i]);
          return EXIT_FAILURE;
        }
        if(header.scalar!=1)
          scalar_set=TRUE;
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[i],argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<i+3)
  {
    fprintf(stderr,"Error: Missing argument(s).\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  if(header.datatype!=LPJ_SHORT && scalar_set)
  {
    fprintf(stderr,"Warning: Scaling set to %g but datatype is %s, scaling set to 1.\n",
            scalar,typenames[header.datatype]);
    header.scalar=1;
    scalar=1;
  }
  rc=nc_open(argv[i],NC_NOWRITE,&ncid);
  if(rc)
  {
    fprintf(stderr,"ERROR409: Cannot open '%s': %s.\n",
            argv[i],nc_strerror(rc));
    return EXIT_FAILURE;
  }
  rc=nc_inq_varid(ncid,var,&var_id);
  if(rc)
  {
    fprintf(stderr,"ERROR405: No variable '%s' found in '%s'.\n",var,argv[i]);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_varndims(ncid,var_id,&ndims);
  if(ndims!=2)
  {
    fprintf(stderr,"ERROR408: Invalid number of dimensions %d in '%s', must be 2.\n",
            ndims,argv[i]);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_vardimid(ncid,var_id,dimids);
  nc_inq_dimname(ncid,dimids[1],name);
  rc=nc_inq_varid(ncid,name,&latlon_id);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read %s in '%s': %s.\n",
            name,argv[i],nc_strerror(rc));
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_dimlen(ncid,dimids[1],&lon_len);
  lon=newvec(double,lon_len);
  if(lon==NULL)
  {
    printallocerr("lon");
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_var_double(ncid,latlon_id,lon);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read longitude in '%s': %s.\n",
            argv[i],nc_strerror(rc));
    free(lon);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_dimname(ncid,dimids[0],name);
  rc=nc_inq_varid(ncid,name,&latlon_id);
  if(rc)
  {
    fprintf(stderr,"ERROR410: Cannot read %s in '%s': %s.\n",
            name,argv[i],nc_strerror(rc));
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  nc_inq_dimlen(ncid,dimids[0],&lat_len);
  if(rc)
  {
    fprintf(stderr,"ERROR403: Cannot read latitude in '%s': %s.\n",
            argv[i],nc_strerror(rc));
    free(lon);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  lat=newvec(double,lat_len);
  if(lat==NULL)
  {
    printallocerr("lat");
    free(lon);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_var_double(ncid,latlon_id,lat);
  if(rc)
  {
    fprintf(stderr,"ERROR404: Cannot read latitude in '%s': %s.\n",
            argv[i],nc_strerror(rc));
    free(lon);
    free(lat);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_att_int(ncid,var_id,"missing_value",&missing_value);
  if(rc)
    rc=nc_get_att_int(ncid,var_id,"_FillValue",&missing_value);
  if(rc)
  {
    fprintf(stderr,"ERROR407: Cannot read missing value in '%s': %s.\n",
            argv[i],nc_strerror(rc));
    free(lon);
    free(lat);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  out=fopen(argv[i+1],"wb");
  if(out==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[i+1],strerror(errno));
    free(lon);
    free(lat);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  fwriteheader(out,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  data=newvec(int,lat_len*lon_len);
  if(data==NULL)
  {
    printallocerr("data");
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  rc=nc_get_var_int(ncid,var_id,data);
  error(rc);
  header.ncell=0;
  soil=fopen(argv[i+2],"wb");
  if(soil==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[i+2],strerror(errno));
    free(lon);
    free(lat);
    free(data);
    nc_close(ncid);
    return EXIT_FAILURE;
  }
  for(ilat=0;ilat<lat_len;ilat++)
  {
    for(ilon=0;ilon<lon_len;ilon++)
    {
      if(data[ilat*lon_len+ilon]!=missing_value)
      {
        coord.lat=lat[ilat];
        coord.lon=lat[ilon];
        writecoord(out,&coord,scalar,header.datatype);
        header.ncell++;
        if(ismap)
          soilcode=(Byte)(soilmap[data[ilat*lon_len+ilon]-1]+1);
        else
          soilcode=(Byte)data[ilat*lon_len+ilon];
        fwrite(&soilcode,1,1,soil);
      }
    }
  }
  free(data);
  fclose(soil);
  rewind(out);
  header.firstcell=0;
  header.nyear=1;
  header.nstep=1;
  header.timestep=1;
  header.firstyear=1901;
  header.nbands=2;
  header.order=CELLYEAR;
  header.cellsize_lon=(lon[lon_len-1]-lon[0])/(lon_len-1);
  header.cellsize_lat=(float)(fabs(lat[lat_len-1]-lat[0])/(lat_len-1));
  if(header.datatype==LPJ_SHORT && (isfloatcoord(header.cellsize_lon*0.5,scalar) || isfloatcoord(header.cellsize_lat*0.5,scalar)))
    fprintf(stderr,"Warning: Cell size (%g,%g) does not allow short datatype for grid with scaling factor %g.\n",
            header.cellsize_lat,header.cellsize_lon,scalar);
  fwriteheader(out,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  fclose(out);
  free(lon);
  free(lat);
  nc_close(ncid);
  printf("Number of cells: %d\n",header.ncell);
  return EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",argv[0]);
  return EXIT_FAILURE;
#endif
} /* of 'main' */
